File: backend/schemas/document.py
from pydantic import BaseModel, Field
from datetime import datetime
from uuid import UUID
from typing import Optional, List
from models.document import DocumentStatus, FileType


class DocumentResponse(BaseModel):
    """Response schema for document data"""
    id: UUID
    filename: str
    file_size: int
    file_type: FileType
    status: DocumentStatus
    created_at: datetime
    processed_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class DocumentUploadResponse(BaseModel):
    """Response schema for document upload"""
    id: UUID
    filename: str
    file_size: int
    file_type: FileType
    status: DocumentStatus
    created_at: datetime
    message: str = "Document uploaded successfully. Processing has started."

    class Config:
        from_attributes = True


class DocumentDetail(BaseModel):
    """Detailed response schema for document with chunk count"""
    id: UUID
    filename: str
    file_size: int
    file_type: FileType
    status: DocumentStatus
    created_at: datetime
    processed_at: Optional[datetime] = None
    chunk_count: int
    error_message: Optional[str] = None

    class Config:
        from_attributes = True


class DocumentList(BaseModel):
    """Response schema for document list"""
    documents: List[DocumentResponse]
    total: int
    limit: int
    offset: int


class DocumentDeleteResponse(BaseModel):
    """Response schema for document deletion"""
    message: str = "Document deleted successfully"
    deleted: dict


class ChunkResponse(BaseModel):
    """Response schema for document text chunks"""
    id: UUID
    document_id: UUID
    chunk_index: int
    text: str
    token_count: int
    created_at: datetime

    class Config:
        from_attributes = True


class DocumentRenameRequest(BaseModel):
    """Request schema for renaming a document"""
    filename: str = Field(..., min_length=1, max_length=255, description="New filename for the document")

File: frontend/types/document.ts
/**
 * Document file type
 */
export type FileType = 'pdf' | 'docx' | 'txt';

/**
 * Document processing status
 */
export type DocumentStatus = 'uploaded' | 'processing' | 'completed' | 'failed';

/**
 * Uploaded document metadata
 */
export interface Document {
  id: string;
  filename: string;
  fileSize: number;
  fileType: FileType;
  status: DocumentStatus;
  createdAt: string;
  processedAt?: string;
  errorMessage?: string;
}

/**
 * Document with chunk count (detailed view)
 */
export interface DocumentDetail extends Document {
  chunkCount: number;
}

/**
 * Text chunk of a document
 */
export interface Chunk {
  id: string;
  documentId: string;
  chunkIndex: number;
  text: string;
  tokenCount: number;
  createdAt: string;
}


/**
 * Document list response with pagination
 */
export interface DocumentListResponse {
  documents: Document[];
  total: number;
  limit: number;
  offset: number;
}

/**
 * Upload progress tracking
 */
export interface UploadProgress {
  filename: string;
  fileSize: number;
  bytesUploaded: number;
  percentage: number;
  status: 'pending' | 'uploading' | 'processing' | 'complete' | 'failed';
  error?: string;
}

/**
 * Usage statistics summary
 */
export interface UsageStats {
  queryCount: number;
  documentCount: number;
  sessionCount: number;
  tokenUsage?: number;
}

/**
 * Activity data point for charts
 */
export interface ActivityDataPoint {
  date: string;
  count: number;
}

/**
 * Document type breakdown
 */
export interface DocumentTypeStats {
  pdf: number;
  docx: number;
  txt: number;
}

/**
 * Analytics dashboard data
 */
export interface AnalyticsDashboard {
  stats: UsageStats;
  activityTimeline: ActivityDataPoint[];
  documentTypes: DocumentTypeStats;
  storageUsed: number;
  storageLimit?: number;
}

/**
 * File upload constraints
 */
export const FILE_UPLOAD_CONSTRAINTS = {
  MAX_SIZE_MB: 10,
  MAX_SIZE_BYTES: 10 * 1024 * 1024,
  ALLOWED_TYPES: [
    'application/pdf',
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    'text/plain'
  ] as const,
  ALLOWED_EXTENSIONS: ['.pdf', '.docx', '.txt'] as const,
} as const;
